### DISCUSSION TITLE
VSCode breakpoints don't work in generated cpp

### DISCUSSION THREAD

**Discussion by mike919192** [hsutter/cppfront#589] [Q&A] [answered]
I won't make a bug report for the project in case this is a VSCode config issue.  Using VSCode with gcc 11 on Ubuntu 22.04.  Using just the simple hello world program, but the breakpoints I set in the generated cpp file don't get hit.  The program itself runs properly.

hello.cpp2
```
main: () -> int = {
    std::cout << "Hello " << name() << "\n";
}

name: () -> std::string = {
    s: std::string = "world";
    decorate(s);
    return s;
}

decorate: (inout s: std::string) = {
    s = "[" + s + "]";
}
```

I can run the binary with gdb through command line and set and hit a breakpoint.  But obviously would prefer to work with the VSCode interface.

And of course a regular old syntax1 cpp file (with no #include "cpp2util.h") I can set and hit breakpoints with VSCode just fine.

Any clues what the issue could be?  Is something in cpp2util.h making VSCode confused?

**Comment by JohelEGP** [comment, marked as answer]
See also <https://github.com/modern-cmake/cppfront/issues/93>.

I have a similar issue with Clang in QtCreator.
When I start debugging with the option "from `main`", which doesn't require a breakpoint, I'm taken to the disassembler.
Like I did, maybe you can work around it by compiling the Cpp2 source file by passing the flag `-clean-cpp1` to `cppfront`.
That makes it so errors at Cpp1 time aren't mapped to the Cpp2 source location, so it helps having another build without `-clean-cpp1`.

**Reply by JohelEGP** [reply]
By the way, have you tried setting the breakpoint on the Cpp2 source file?
I remember that working with GCC in QtCreator.

**Reply by mike919192** [reply]
Thanks!  It was indeed the `#line 2 "hello.cpp2"` statements that we causing the breakpoints to be disabled.  Compiling the cpp2 file with the -c flag produced a cpp that I could hit a breakpoint.  I'll have to do some searching to see why VSCode behaves this way.

I haven't been brave enough to try to setup a VSCode configuration for cpp2 files.  One step at a time :)

**Reply by JohelEGP** [reply]
Good luck! And please, do share the results of your research.

**Reply by mike919192** [reply]
OK I've got things working with the line directives now.  So it seems that when debugging a file with line directives, VSCode expects you to place breakpoints in the file that the statement is specifying, so the cpp2 file in this case.  It even goes so far as to disable any breakpoints in the generated cpp file.

Also by default, VSCode wont let you place breakpoints in the cpp2 file because it doesn't know the extension.  So I added a file association for cpp2 and h2 to cpp.
![image](https://github.com/hsutter/cppfront/assets/91038685/cfc73e75-55dd-460b-83cc-e5e38bd5bb0f)

Now I can set breakpoints in the cpp2 and h2 files, then debug the cpp file and the breakpoints will be hit.  Stepping around and looking at variables all seems to work from the cpp2 / h2 files.  Of course intellisense doesnt like the cpp2 file because it doesn't know the format but that doesn't prevent debugging.
